src/contracts.rs

```diff
@@ -58,132 +64,67 @@ fn only_ignored_fields<V>(extensions: &BTreeMap<String, V>) -> bool {
+    /// converted to a (lazy) contract, and thus requires to go through a predicate.
+    /// Upon success, returns the contract and the references used in the schema.
+    pub fn from_root_schema(root_schema: &RootSchema) -> Option<(Self, RefsUsage)> {
+        let mut refs_usage = RefsUsage::new();
 
-impl From<Contract> for RichTerm {
-    fn from(Contract(c): Contract) -> Self {
-        match c.as_slice() {
-            [] => Predicate::from(&Schema::Bool(true)).into(),
-            // TODO: shouldn't need to clone here
-            [rt] => rt.clone(),
-            _ => {
-                let arr = Term::Array(Array::new(c.into_iter().collect()), Default::default());
-                mk_app!(static_access("std", ["contract", "Sequence"]), arr)
-            }
-        }
+        root_schema
+            .schema
+            .try_as_contract(&mut refs_usage)
+            .map(|ctr| (ctr, refs_usage))
     }
-}
 
-impl From<Term> for Contract {
-    fn from(value: Term) -> Self {
-        Contract::from(RichTerm::from(value))
+    /// Return the `Dyn` contract, always succeeding.
+    pub fn dynamic() -> Self {
+        Term::Type(TypeF::Dyn.into()).into()
     }
 }
 
-impl From<TypeF<Box<Type>, RecordRows, EnumRows>> for Contract {
-    fn from(value: TypeF<Box<Type>, RecordRows, EnumRows>) -> Self {
-        Contract::from(Term::Type(Type::from(value)))
-    }
+/// [TryAsContract] is essentially like `TryInto<Contract>` but passes additional state around used for
+/// effective reference resolution.
+pub trait TryAsContract {
+    /// Try to convert a JSON schema component `Self` to a contract. Returns `None` if the
+    /// component couldn't be converted to a lazy contract, and thus requires to go through a
+    /// predicate.
+    ///
+    /// `try_as_contract` will record the references used during the conversion through the `refs_usage` parameter.
+    fn try_as_contract(&self, refs_usage: &mut RefsUsage) -> Option<Contract>;
 }
 
-impl From<&InstanceType> for Contract {
-    fn from(value: &InstanceType) -> Contract {
-        match value {
-            InstanceType::Null => contract_from_predicate(
-                mk_app!(
-                    static_access("predicates", ["isType"]),
-                    Term::Enum("Null".into())
-                )
-                .into(),
-            ),
-            InstanceType::Boolean => Contract::from(TypeF::Bool),
-            InstanceType::Object => Contract::from(Term::Record(RecordData {
-                attrs: RecordAttrs { open: true },
-                ..Default::default()
-            })),
-            InstanceType::Array => Contract::from(TypeF::Array(Box::new(TypeF::Dyn.into()))),
-            InstanceType::Number => Contract::from(TypeF::Number),
-            InstanceType::String => Contract::from(TypeF::String),
-            InstanceType::Integer => Contract::from(static_access("std", ["number", "Integer"])),
-        }
-    }
+pub trait AsCtrThruPred {
```

Comment:
I understand that `AsContractThroughPredicate` feels very Java, but `AsCtrThruPred` doesn't exactly spark joy either. I think with @Radvendii we got used to calling contracts that resort to going through the predicate "predicate contracts". Maybe `AsPredicateContract` would be a reasonable name?

Reply:
Yeah, it's a fine balance and it's hard to find. I think Rust in general favor shorter names with missing letters. I think I (and you, and we in general) favor longer, more explicit names. However I think it got a bit out of hand in e.g. the Nickel typechecker (yes, I'm thinking of you, `GenericUnifRecordRowsIteratorItem`), and I came to think that having abbreviation over a certain limit is ok as long as the meaning of the abbreviation is absolutely non ambiguous. A bit like comments: comments are good, but they need to have some entropy, and paraphrasing the code is usually just noise.

Anyway, I think the `PredicateContract` is a good alternative. Another would be `EagerContract`. We haven't used the latter yet though, so maybe `PredicateContract` is better, although it might be a tad confusing?